client/network-gossip/src/bridge.rs

```diff
@@ -0,0 +1,299 @@
+		state_machine.register_validator(&mut *context, engine_id, validator);
+
+		let inner = Arc::new(Mutex::new(GossipEngineInner {
+			state_machine,
+			context,
+			context_ext,
+		}));
+
+		let gossip_engine = GossipEngine {
+			inner: inner.clone(),
+			engine_id,
+		};
+
+		let res = executor.spawn({
+			let inner = Arc::downgrade(&inner);
+			async move {
+				loop {
+					let _ = futures_timer::Delay::new(Duration::from_millis(1100)).await;
+					if let Some(inner) = inner.upgrade() {
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						inner.state_machine.tick(&mut *inner.context);
+					} else {
+						break;
+					}
+				}
+			}
+		});
+
+		// Note: we consider the chances of an error to spawn a background task almost null.
+		if res.is_err() {
+			log::error!(target: "gossip", "Failed to spawn background task");
+		}
+
+		let res = executor.spawn(async move {
+			let mut stream = Compat01As03::new(event_stream);
+			while let Some(Ok(event)) = stream.next().await {
+				match event {
+					Event::NotifOpened { remote, engine_id: msg_engine_id } => {
+						if msg_engine_id != engine_id {
+							continue;
+						}
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						// TODO: for now we hard-code the roles to FULL; fix that
+						inner.state_machine.new_peer(&mut *inner.context, remote, Roles::FULL);
+					}
+					Event::NotifClosed { remote, engine_id: msg_engine_id } => {
+						if msg_engine_id != engine_id {
+							continue;
+						}
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						inner.state_machine.peer_disconnected(&mut *inner.context, remote);
+					},
+					Event::NotifMessages { remote, messages } => {
+						let mut inner = inner.lock();
+						let inner = &mut *inner;
+						inner.state_machine.on_incoming(
+							&mut *inner.context,
+							remote,
+							messages.into_iter()
+								.filter_map(|(engine, data)| if engine == engine_id {
+									Some(ConsensusMessage { engine_id: engine, data: data.to_vec() })
+								} else { None })
+								.collect()
+						);
+					},
+					Event::Dht(_) => {}
+				}
+			}
+		});
+
+		// Note: we consider the chances of an error to spawn a background task almost null.
+		if res.is_err() {
+			log::error!(target: "gossip", "Failed to spawn background task");
+		}
+
+		gossip_engine
+	}
+
+	/// Closes all notification streams.
+	pub fn abort(&self) {
+		self.inner.lock().state_machine.abort();
+	}
+
+	pub fn report(&self, who: PeerId, reputation: ReputationChange) {
+		self.inner.lock().context.report_peer(who, reputation);
+	}
+
+	/// Registers a message without propagating it to any peers. The message
+	/// becomes available to new peers or when the service is asked to GossipEngine
+	/// the message's topic. No validation is performed on the message, if the
+	/// message is already expired it should be dropped on the next garbage
+	/// collection.
+	pub fn register_gossip_message(
```

Comment:
Why would we register a message without propagating to existing peers, but send it to new peers? Is this used somewhere?

Reply:
That was already there before my refactoring.